File: commands/match.js
import { SlashCommandBuilder, EmbedBuilder, MessageFlags } from 'discord.js';
import { readTournamentData, readMatchVotes } from '../utils/firebase.js';
import { getMatchStake } from '../utils/football.js';
import { getWinner } from '../utils/helper.js';
import { withErrorHandler, getTournamentName } from '../utils/command.js';

export const data = new SlashCommandBuilder()
  .setName('match')
  .setDescription('Chi tiết trận đấu — ai đúng ai sai, rõ ràng')
  .addIntegerOption(option => option.setName('id')
    .setDescription('ID trận (bỏ trống = trận gần nhất)')
    .setMinValue(1)
    .setRequired(false));

export const execute = withErrorHandler(async (interaction) => {
  const tournamentName = await getTournamentName();
  const allMatches = await readTournamentData('matches');

  if (!allMatches) {
    await interaction.reply({ content: '❌ Không có dữ liệu trận đấu.', flags: MessageFlags.Ephemeral });
    return;
  }

  const matchIdOption = interaction.options.get('id')?.value;
  let match;
  if (matchIdOption) {
    match = allMatches.find((m) => m.id === matchIdOption);
  } else {
    match = allMatches
      .filter((m) => m.hasResult)
      .sort((a, b) => Date.parse(b.date) - Date.parse(a.date))[0];
  }

  if (!match) {
    const embed = new EmbedBuilder()
      .setTitle('❌  Không tìm thấy trận')
      .setDescription(matchIdOption ? `Không có trận nào ID \`${matchIdOption}\` cả.` : 'Chưa có trận nào xong.')
      .setColor(0xED4245);
    await interaction.reply({ embeds: [embed], flags: MessageFlags.Ephemeral });
    return;
  }

  const kickoff = new Date(match.date);
  const ts = Math.floor(kickoff.getTime() / 1000);
  const now = Date.now();
  const hasStarted = kickoff.getTime() <= now;

  let status;
  if (match.hasResult) {
    status = `✅ Kết thúc — **${match.home.toUpperCase()}** ${match.result.home} - ${match.result.away} **${match.away.toUpperCase()}**`;
  } else if (hasStarted) {
    status = '🔴 Đang đá / chờ kết quả';
  } else {
    status = `🟢 Sắp đá — <t:${ts}:R>`;
  }

  const stake = getMatchStake(match.id);
  const embed = new EmbedBuilder()
    .setTitle(`⚽  Match #${match.id}: ${match.home.toUpperCase()} vs ${match.away.toUpperCase()}`)
    .setDescription(`**${tournamentName}**\n\n${status}`)
    .setColor(match.hasResult ? 0x57F287 : hasStarted ? 0xED4245 : 0x5865F2)
    .addFields(
      { name: '🕐 Giờ đá', value: `<t:${ts}:f>`, inline: true },
      { name: '🏟️ Sân', value: match.location, inline: true },
      { name: '💰 Cược', value: `${stake} pts`, inline: true },
    )
    .setTimestamp();

  if (hasStarted && match.messageId) {
    const votes = await readMatchVotes(match.id, match.messageId);
    if (votes) {
      const users = interaction.client.cachedUsers;
      const winner = match.hasResult ? getWinner(match) : null;
      const grouped = {};
      for (const [userId, v] of Object.entries(votes)) {
        const pick = v.vote.toUpperCase();
        if (!grouped[pick]) grouped[pick] = [];
        const name = users[userId]?.nickname || 'Unknown';
        const icon = winner ? (v.vote === winner ? '👑' : '🤡') : '🗳️';
        grouped[pick].push(`${icon} ${name}`);
      }
      let voteLines = Object.entries(grouped)
        .map(([pick, names]) => `**${pick}**\n${names.join('\n')}`)
        .join('\n\n');
      if (voteLines.length > 1024) voteLines = voteLines.slice(0, 1021) + '...';
      embed.addFields({ name: '🗳️ Vote', value: voteLines, inline: false });
    } else {
      embed.addFields({ name: '🗳️ Vote', value: '*Không có vote*', inline: false });
    }
  }

  await interaction.reply({ embeds: [embed] });
});

File: commands/match.go
package commands

import (
	"context"
	"fmt"
	"sort"
	"time"

	"github.com/kdnvi/rambo-bot/bot"
	"github.com/kdnvi/rambo-bot/internal/discord"
	"github.com/kdnvi/rambo-bot/internal/firebase"
)

func Match(ctx context.Context, b *bot.Bot, i *discord.Interaction) {
	tournamentName := getTournamentName(ctx, b)

	matches, err := b.DB.ReadMatches(ctx)
	if err != nil || matches == nil {
		replyErr(ctx, b, i, "❌ Không có dữ liệu trận đấu.")
		return
	}

	var match *firebase.Match
	if opt := i.Data.GetOption("id"); opt != nil {
		id := int(opt.Int())
		for idx := range matches {
			if matches[idx].ID == id {
				match = &matches[idx]
				break
			}
		}
		if match == nil {
			replyEphemeral(ctx, b, i, discord.InteractionCallbackData{
				Embeds: []discord.Embed{{
					Title:       "❌  Không tìm thấy trận",
					Description: fmt.Sprintf("Không có trận nào ID `%d` cả.", id),
					Color:       0xED4245,
				}},
			})
			return
		}
	} else {
		// Most recent completed match
		sort.Slice(matches, func(a, b int) bool {
			ta, _ := time.Parse(time.RFC3339, matches[a].Date)
			tb, _ := time.Parse(time.RFC3339, matches[b].Date)
			return ta.After(tb)
		})
		for idx := range matches {
			if matches[idx].HasResult {
				match = &matches[idx]
				break
			}
		}
		if match == nil {
			replyEphemeral(ctx, b, i, discord.InteractionCallbackData{
				Embeds: []discord.Embed{{
					Title:       "❌  Không tìm thấy trận",
					Description: "Chưa có trận nào xong.",
					Color:       0xED4245,
				}},
			})
			return
		}
	}

	kickoff, _ := time.Parse(time.RFC3339, match.Date)
	ts := kickoff.Unix()
	now := time.Now()
	hasStarted := kickoff.Before(now)

	var status string
	var color int
	switch {
	case match.HasResult:
		status = fmt.Sprintf("✅ Kết thúc — **%s** %d - %d **%s**",
			upper(match.Home), match.Result.Home, match.Result.Away, upper(match.Away))
		color = 0x57F287
	case hasStarted:
		status = "🔴 Đang đá / chờ kết quả"
		color = 0xED4245
	default:
		status = fmt.Sprintf("🟢 Sắp đá — <t:%d:R>", ts)
		color = 0x5865F2
	}

	stake := getMatchStake(match.ID)
	embed := discord.Embed{
		Title:       fmt.Sprintf("⚽  Match #%d: %s vs %s", match.ID, upper(match.Home), upper(match.Away)),
		Description: fmt.Sprintf("**%s**\n\n%s", tournamentName, status),
		Color:       color,
		Fields: []discord.EmbedField{
			{Name: "🕐 Giờ đá", Value: fmt.Sprintf("<t:%d:f>", ts), Inline: true},
			{Name: "🏟️ Sân", Value: match.Location, Inline: true},
			{Name: "💰 Cược", Value: fmt.Sprintf("%d pts", stake), Inline: true},
		},
		Timestamp: time.Now().UTC().Format(time.RFC3339),
	}

	if hasStarted && match.MessageID != "" {
		votes, err := b.DB.ReadMatchVotes(ctx, match.ID, match.MessageID)
		if err == nil && votes != nil {
			users := b.GetAllCachedUsers()
			winner := getWinner(match)
			grouped := map[string][]string{}
			for uid, v := range votes {
				pick := upper(v.Vote)
				icon := "🗳️"
				if winner != "" {
					if v.Vote == winner {
						icon = "👑"
					} else {
						icon = "🤡"
					}
				}
				name := "Unknown"
				if u := users[uid]; u != nil {
					name = u.DisplayName()
				}
				grouped[pick] = append(grouped[pick], icon+" "+name)
			}
			voteLines := ""
			for pick, names := range grouped {
				voteLines += fmt.Sprintf("**%s**\n%s\n\n", pick, joinLines(names, "\n"))
			}
			if len(voteLines) > 1024 {
				voteLines = voteLines[:1021] + "..."
			}
			if voteLines == "" {
				voteLines = "*Không có vote*"
			}
			embed.Fields = append(embed.Fields, discord.EmbedField{Name: "🗳️ Vote", Value: voteLines})
		} else {
			embed.Fields = append(embed.Fields, discord.EmbedField{Name: "🗳️ Vote", Value: "*Không có vote*"})
		}
	}

	reply(ctx, b, i, discord.InteractionCallbackData{Embeds: []discord.Embed{embed}})
}
